File: 0322-coin-change/0322-coin-change.cpp
class Solution {
public:
    int coinChange(vector<int>& coins, int amount) {
        vector<vector<int>> dp (coins.size(), vector<int> (amount + 1, -1));

        int result = recursion(coins.size() - 1, amount, coins, dp);
        if (result >= 1e9) return -1;
        return result;
    }

    int recursion(int idx, int amount, vector<int>& coins, vector<vector<int>>& dp){

        if (amount == 0) return 0;
        if (idx == 0) {
            if (amount % coins[0] == 0) return amount/coins[0];
            return 1e9;
        }

        if (dp[idx][amount] != -1) return dp[idx][amount];

        int notTaken = recursion(idx - 1, amount, coins, dp);
        int taken = 1e9;
        if (amount >= coins[idx]){
            taken = 1 + recursion(idx, amount - coins[idx], coins, dp);
        }


        
        return dp[idx][amount] = min(taken, notTaken);
    }
};

File: 0322-coin-change/0322-coin-change.java
class Solution {
    int dp[][];

    public int coinChange(int[] coins, int amount) {
        int n = coins.length;
        int dp[][] = new int[n + 1][amount + 1];

        for (int i = 0; i <= amount; i++) {
            dp[0][i] = Integer.MAX_VALUE;
        }    
        dp[0][0] = 0;

        for (int i = 1; i <= n; i++) {
            for (int amt = 0; amt <= amount; amt++) {
                int take = Integer.MAX_VALUE;

                if (coins[i - 1] <= amt) {
                    int result = dp[i][amt - coins[i - 1]] ;
                    if (result != Integer.MAX_VALUE) {
                        take = result + 1;
                    }
                } 
                
                int notTake = dp[i - 1][amt];
                

                dp[i][amt] = Math.min(take, notTake);
            } 
        }

        return dp[n][amount] == Integer.MAX_VALUE? -1: dp[n][amount];
    }


}